### PR TITLE
[core] Add cluster ID to the Python layer

### PR DESCRIPTION
**Pull request by vitsai**
## Why are these changes needed?

An earlier[ change](https://github.com/ray-project/ray/pull/37399) added a new RPC to each GCS client, causing some stress tests to fail. Instead of adding O(num_worker) RPCs, add O(num_node) RPCs by plumbing through the Python layer.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/raylet/main.cc [hunk 1]
```diff
@@ -149,6 +151,11 @@ int main(int argc, char *argv[]) {
   const std::string session_name = FLAGS_session_name;
   const bool is_head_node = FLAGS_head;
   const std::string labels_json_str = FLAGS_labels;
+
+  ray::ClusterID cluster_id = ray::ClusterID::Nil();
+  if (FLAGS_cluster_id != "") {
+    cluster_id = ray::ClusterID::FromHex(FLAGS_cluster_id);
+  }
```
Comment (iycheng) [inline src/ray/raylet/main.cc:158]: Print some logs here for future debugging purpose.

File: src/ray/gcs/gcs_client/gcs_client.h [hunk 2]
```diff
@@ -241,6 +241,19 @@ class RAY_EXPORT PythonGcsClient {
...
+    grpc::ClientContext context;
+    if (timeout_ms != -1) {
+      context.set_deadline(std::chrono::system_clock::now() +
+                           std::chrono::milliseconds(timeout_ms));
+    }
+    if (!cluster_id_.IsNil()) {
+      context.AddMetadata(kClusterIdKey, cluster_id_.Hex());
+    }
+    return std::move(context);
+  }
```
Comment (iycheng) [inline src/ray/gcs/gcs_client/gcs_client.h:254]:
```suggestion
 grpc::ClientContext PrepareContext(int64_t timeout_ms) {
    grpc::ClientContext context;
    if (timeout_ms != -1) {
      context.set_deadline(std::chrono::system_clock::now() +
                           std::chrono::milliseconds(timeout_ms));
    }
    if (!cluster_id_.IsNil()) {
      context.AddMetadata(kClusterIdKey, cluster_id_.Hex());
    }
    return context;
  }
```

File: src/ray/gcs/gcs_client/gcs_client.cc [hunk 3]
```diff
@@ -432,8 +416,7 @@ Status PythonGcsClient::GetAllJobInfo(int64_t timeout_ms,
 Status PythonGcsClient::RequestClusterResourceConstraint(
     int64_t timeout_ms,
     const std::vector<std::unordered_map<std::string, double>> &bundles) {
-  grpc::ClientContext context;
-  GrpcClientContextWithTimeoutMs(context, timeout_ms);
+  grpc::ClientContext &&context = PrepareContext(timeout_ms);
```
Comment (iycheng) [inline src/ray/gcs/gcs_client/gcs_client.cc:419]:
```suggestion
  grpc::ClientContext context = PrepareContext(timeout_ms);
```

File: python/ray/includes/common.pxd [hunk 4]
```diff
@@ -334,8 +335,9 @@ cdef extern from "ray/gcs/gcs_client/gcs_client.h" nogil:
     cdef cppclass CPythonGcsClient "ray::gcs::PythonGcsClient":
         CPythonGcsClient(const CGcsClientOptions &options)
 
-        CRayStatus Connect()
-
+        CRayStatus Connect(
+            CClusterID &cluster_id,
```
Comment (iycheng) [inline python/ray/includes/common.pxd:339]: const?

File: src/ray/gcs/gcs_client/gcs_client.cc [hunk 5]
```diff
@@ -177,7 +146,44 @@ std::pair<std::string, int> GcsClient::GetGcsServerAddress() const {
...
-Status PythonGcsClient::Connect() {
+namespace {
+Status HandleGcsError(rpc::GcsStatus status) {
+  RAY_CHECK(status.code() != static_cast<int>(StatusCode::OK));
+  return Status::Invalid(status.message() +
+                         " [GCS status code: " + std::to_string(status.code()) + "]");
+}
+}  // namespace
+
+Status PythonGcsClient::Connect(ClusterID &cluster_id, int64_t timeout_ms) {
```
Comment (iycheng) [inline src/ray/gcs/gcs_client/gcs_client.cc:157]: The reference of cluster_id makes it an in and out parameter is confusing.

Maybe let's add const there. Keep a variable inside the class as cluster_id. Give a getter function.

**Review by fishbone — COMMENTED** [review]
Overall it looks good.

**Comment by iycheng** [comment]
Could you run the failed release test to make sure it's working?

File: python/ray/_private/node.py [hunk 6]
```diff
@@ -1014,6 +1018,7 @@ def start_raylet(
             self._ray_params.node_manager_port,
             self._raylet_socket_name,
             self._plasma_store_socket_name,
+            self._cluster_id,
```
Comment (jjyao) [inline python/ray/_private/node.py:1021]: you mean `self.cluster_id`?

File: python/ray/_private/worker.py [hunk 7]
```diff
@@ -461,6 +461,7 @@ def __init__(self):
         # different drivers that connect to the same Serve instance.
         # See https://github.com/ray-project/ray/pull/35070.
         self._filter_logs_by_job = True
+        self.cluster_id = None
```
Comment (jjyao) [inline python/ray/_private/worker.py:464]: Is this used?

File: src/ray/gcs/gcs_client/gcs_client.cc [hunk 8]
```diff
@@ -177,7 +146,44 @@ std::pair<std::string, int> GcsClient::GetGcsServerAddress() const {
...
+  if (cluster_id.IsNil()) {
+    RAY_LOG(INFO) << "Retrieving cluster ID from GCS server.";
+    grpc::ClientContext &&context = PrepareContext(timeout_ms);
+
+    rpc::GetClusterIdRequest request;
+    rpc::GetClusterIdReply reply;
+
+    auto status =
+        GrpcStatusToRayStatus(node_info_stub_->GetClusterId(&context, request, &reply));
```
Comment (jjyao) [inline src/ray/gcs/gcs_client/gcs_client.cc:166]: I'm confused, isn't `node_info_stub_` created in line 191 below?

File: src/ray/gcs/gcs_client/gcs_client.h (hunk 2 above)
Comment (vitsai) [reply src/ray/gcs/gcs_client/gcs_client.h]: Compiler complains because there's no copy constructor for ClientContext. Moved the ClientContext out of the helper the way it was before.

File: python/ray/_private/node.py [hunk 9]
```diff
@@ -83,6 +83,9 @@ def __init__(
         )
         self.all_processes: dict = {}
         self.removal_lock = threading.Lock()
+        self.cluster_id = (
+            ray_params.cluster_id if hasattr(ray_params, "cluster_id") else None
```
[the remark this answers is not in the document]
Comment (vitsai) [reply python/ray/_private/node.py]: Originally it was because there was no enforcement that the attr was even declared, and when it's not declared, it throws exception instead of returning None (i.e. if `ray_params.cluster_id = cluster_id` is only called sometimes). 

But RayParams is fixed now per your other comment.

**Review by vitsai — COMMENTED** [review]
good comments

File: src/ray/gcs/gcs_client/gcs_client.h (hunk 2 above)
Comment (iycheng) [reply src/ray/gcs/gcs_client/gcs_client.h]: hmmm, interesting... because you actually can return a unique_ptr directly.

File: src/ray/gcs/gcs_client/gcs_client.h [hunk 10]
```diff
@@ -231,7 +232,20 @@ class RAY_EXPORT PythonGcsClient {
       const std::vector<std::unordered_map<std::string, double>> &bundles);
   Status GetClusterStatus(int64_t timeout_ms, std::string &serialized_reply);
 
+  ClusterID GetClusterId() const { return cluster_id_; }
```
Comment (iycheng) [inline src/ray/gcs/gcs_client/gcs_client.h:235]: nit: `const ClusterID&` maybe? not big deal btw.

**Review by fishbone — APPROVED** [review]
Thanks! Let's make sure the broken release test passes

File: python/ray/_private/node.py [hunk 11]
```diff
@@ -83,6 +83,7 @@ def __init__(
         )
         self.all_processes: dict = {}
         self.removal_lock = threading.Lock()
+        self.cluster_id = ray_params.cluster_id
```
Comment (jjyao) [inline python/ray/_private/node.py:86]: I'd suggest to remove this line and only set it inside `_init_gcs_client`. This way we make sure `self.cluster_id` can never be None.

File: python/ray/_private/node.py [hunk 12]
```diff
@@ -598,7 +599,8 @@ def _init_gcs_client(self):
             last_ex = None
             try:
                 gcs_address = self.gcs_address
-                client = GcsClient(address=gcs_address)
+                client = GcsClient(address=gcs_address, cluster_id=self.cluster_id)
```
Comment (jjyao) [inline python/ray/_private/node.py:602]:
```suggestion
                client = GcsClient(address=gcs_address, cluster_id=self._ray_params.cluster_id)
```

File: python/ray/_private/parameter.py [hunk 13]
```diff
@@ -185,6 +185,7 @@ def __init__(
         env_vars: Optional[Dict[str, str]] = None,
         session_name: Optional[str] = None,
         webui: Optional[str] = None,
+        cluster_id: Optional[str] = None,
```
Comment (jjyao) [inline python/ray/_private/parameter.py:188]: Please add a docstring for it above.

File: python/ray/includes/unique_ids.pxd [hunk 14]
```diff
@@ -154,6 +157,16 @@ cdef extern from "ray/common/id.h" namespace "ray" nogil:
...
 
+    cdef cppclass CClusterID "ray::ClusterID"(CUniqueID):
+
+        @staticmethod
+        CClusterID FromHex(const c_string &hex_str)
+
+        @staticmethod
+        const CClusterID Nil()
+
+#        c_bool IsNil() const
```
Comment (jjyao) [inline python/ray/includes/unique_ids.pxd:168]: ?

File: src/ray/gcs/gcs_client/gcs_client.cc [hunk 15]
```diff
@@ -145,36 +146,61 @@ std::pair<std::string, int> GcsClient::GetGcsServerAddress() const {
 
 PythonGcsClient::PythonGcsClient(const GcsClientOptions &options) : options_(options) {}
 
-Status PythonGcsClient::Connect() {
+namespace {
+Status HandleGcsError(rpc::GcsStatus status) {
+  RAY_CHECK(status.code() != static_cast<int>(StatusCode::OK));
```
Comment (jjyao) [inline src/ray/gcs/gcs_client/gcs_client.cc:151]: RAY_CHECK_NE

File: src/ray/gcs/gcs_client/gcs_client.cc [hunk 16]
```diff
@@ -145,36 +146,61 @@ std::pair<std::string, int> GcsClient::GetGcsServerAddress() const {
...
+    RAY_LOG(INFO) << "Retrieving cluster ID from GCS server.";
+    grpc::ClientContext context;
+    PrepareContext(context, timeout_ms);
+
+    rpc::GetClusterIdRequest request;
+    rpc::GetClusterIdReply reply;
+
+    auto status =
+        GrpcStatusToRayStatus(node_info_stub_->GetClusterId(&context, request, &reply));
+    while (!status.IsTimedOut()) {
```
Comment (jjyao) [inline src/ray/gcs/gcs_client/gcs_client.cc:176]: Should we fail after several retries?

File: src/ray/raylet/main.cc [hunk 17]
```diff
@@ -149,6 +151,10 @@ int main(int argc, char *argv[]) {
   const std::string session_name = FLAGS_session_name;
   const bool is_head_node = FLAGS_head;
   const std::string labels_json_str = FLAGS_labels;
+
+  RAY_CHECK(FLAGS_cluster_id != "") << "Expected cluster ID.";
```
Comment (jjyao) [inline src/ray/raylet/main.cc:155]: RAY_CHECK_NE

**Review by jjyao — COMMENTED** [review]
LGTM

File: src/ray/gcs/gcs_client/gcs_client.cc (hunk 16 above)
Comment (vitsai) [reply src/ray/gcs/gcs_client/gcs_client.cc:176]: Looks like all the other methods use timeouts

File: python/ray/_private/node.py [hunk 18]
```diff
@@ -1014,6 +1017,7 @@ def start_raylet(
             self._ray_params.node_manager_port,
             self._raylet_socket_name,
             self._plasma_store_socket_name,
+            self.cluster_id.decode(),
```
Comment (jjyao) [inline python/ray/_private/node.py:1020]: Why do we need to `decode()`? Is `self.cluster_id` the hex string?

**Review by jjyao — APPROVED** [review]

File: python/ray/_private/node.py (hunk 18 above)
Comment (vitsai) [reply python/ray/_private/node.py:1020]: Yes :) At some point it was getting serialized with an extra 'b' in front

Comment (jjyao) [reply python/ray/_private/node.py:1020]: let's make sure `get_cluster_id()` returns the decoded hex string?

File: src/ray/gcs/gcs_client/gcs_client.cc (hunk 16 above)
Comment (vitsai) [reply src/ray/gcs/gcs_client/gcs_client.cc]: Added retries to emulate the @_auto_reconnect on the python side

File: dashboard/modules/job/tests/test_job_agent.py [hunk 19]
```diff
@@ -518,8 +518,14 @@ def test_agent_logs_not_streamed_to_drivers():
     err_str = proc.stderr.read().decode("ascii")
 
     print(out_str, err_str)
-    assert "(raylet)" not in out_str
-    assert "(raylet)" not in err_str
+
+    def check_output(blob):
```
Comment (iycheng) [inline dashboard/modules/job/tests/test_job_agent.py:522]: Not quite sure, but the test is to make sure no logs are streamed to worker from raylet. SHould we keep that?

File: python/ray/_private/worker.py [hunk 20]
```diff
@@ -1126,6 +1126,7 @@ def init(
     namespace: Optional[str] = None,
     runtime_env: Optional[Union[Dict[str, Any], "RuntimeEnv"]] = None,  # noqa: F821
     storage: Optional[str] = None,
+    no_gcs: bool = False,
```
Comment (iycheng) [inline python/ray/_private/worker.py:1129]: We shouldn't change the API.

**Review by fishbone — CHANGES_REQUESTED** [review]
We shouldn't change the API. Still in progress of reviewing.

File: python/ray/_private/worker.py (hunk 20 above)
Comment (vitsai) [reply python/ray/_private/worker.py:1129]: should we have a global variable then? initialization order might be tricky

Comment (vitsai) [reply python/ray/_private/worker.py:1129]: So the problem is that some tests don't start GCS, and then the newly added RPC hangs. Also, monitor uses this for now as it seems GCS is not always available to it.

File: dashboard/modules/job/tests/test_job_agent.py (hunk 19 above)
Comment (vitsai) [reply dashboard/modules/job/tests/test_job_agent.py:522]: But we added new logs in this change that aren't errors, so we broke that invariant. Probably doesn't make sense to test for it anymore

File: python/ray/_private/node.py [hunk 21]
```diff
@@ -649,19 +654,26 @@ def _init_gcs_client(self):
...
-                errors = [e for e in err.readlines() if " C " in e or " E " in e][-10:]
-            error_msg = "\n" + "".join(errors) + "\n"
-            raise RuntimeError(
-                f"Failed to {'start' if self.head else 'connect to'} GCS. "
-                f" Last {len(errors)} lines of error files:"
-                f"{error_msg}."
-                f"Please check {os.path.join(self._logs_dir, 'gcs_server.out')}"
-                " for details"
-            )
+            if hasattr(self, "_logs_dir"):
```
[the remark this answers is not in the document]
Comment (vitsai) [reply python/ray/_private/node.py:657]: I think this was always a bug, but the path never got triggered. In the init path `_logs_dir` is inited after GcsClient

File: dashboard/modules/job/tests/test_job_agent.py (hunk 19 above)
Comment (vitsai) [reply dashboard/modules/job/tests/test_job_agent.py]: ok fixed

File: python/ray/tests/test_ray_init.py [hunk 22]
```diff
@@ -125,7 +125,7 @@ def test_ray_init_existing_instance_crashed(address):
         # If no address is specified, we will default to an existing cluster.
         ray._private.node.NUM_REDIS_GET_RETRIES = 1
         with pytest.raises(ConnectionError):
-            ray.init(address=address)
+            ray.init(address=address, no_gcs=True)
```
Comment (jjyao) [inline python/ray/tests/test_ray_init.py:128]: What's this?

File: python/ray/tests/test_autoscaler.py [hunk 23]
```diff
@@ -3514,14 +3514,18 @@ class FaultyAutoscaler:
             def __init__(self, *args, **kwargs):
                 raise AutoscalerInitFailException
 
+        os.environ["RAY_GCS_SERVER_PORT"] = "12345"
+        ray.init()
```
Comment (iycheng) [inline python/ray/tests/test_autoscaler.py:3518]: consider using fixture here. ray init won't be clean up after this test.

(or at least with manager)

File: python/ray/tests/test_tqdm.py [hunk 24]
```diff
@@ -35,9 +35,9 @@ def update(self):
     assert not mgr.in_hidden_state
 
     # Test stdout save/restore clearing.
-    assert mgr.num_hides == 0
+    num_hides = mgr.num_hides
```
Comment (iycheng) [inline python/ray/tests/test_tqdm.py:38]: why update this? seems not related to this pr

File: python/ray/train/BUILD [hunk 25]
```diff
@@ -342,6 +342,14 @@ py_test(
     deps = [":train_lib", ":conftest"]
 )
 
+py_test(
+    name = "test_gpu_2",
+    size = "large",
```
Comment (iycheng) [inline python/ray/train/BUILD:347]: should it be a large?

**Review by fishbone — APPROVED** [review]
Please fix the comment and run the release tests (many actors/tasks/dead actors) to ensure no regression.

**Comment by vitsai** [comment]
Here are release tests from last night: https://buildkite.com/ray-project/release-tests-pr/builds/47573

Since the comments are nits, I propose we don't rerun the release tests for the latest version.

File: python/ray/tests/test_autoscaler.py (hunk 23 above)
Comment (vitsai) [reply python/ray/tests/test_autoscaler.py:3518]: Apparently pytest fixtures don't work with the unittest package, and the usual way to do this seems to be in the setup and teardown methods. Since this is specific to this particular method, I just unset it afterwards. We can clean up the test in a followup.

File: python/ray/train/BUILD (hunk 25 above)
Comment (vitsai) [reply python/ray/train/BUILD]: looks like it ran on CI in 95 seconds, so changed it to medium.
